### ISSUE TITLE
NVivo import

### ISSUE THREAD

**Issue by glocalglocal** [ccbogel/QualCoder#106]
Better than I thought. All texts are there, they are coded but there are issues similar to those we encountered when importing from RQDA. 

A small number of special characters didn't make it -- eg apostrophes and quotes

It is clear texts are coded but the Code Text dialogue shows the wrong colours in the tree (lots of greys and magentas but all codes are visible) and no colours in the texts until I change the coder name. Shouldn't 'Show other coders' (always greyed out) solve this? When I change the coder name I get the codings colours but then when I hover over the codings I get: 

2020/07/09 00:56:48 ERROR select_file.exception_handler Uncaught exception: Traceback (most recent call last):
  File "C:\Other programs\QualCoder-master\qualcoder\code_text.py", line 1164, in eventFilter
    displayText = '<p style="background-color:' + item['color'] + '"><em>' + item['name'] + "</em><br />" + seltext + "</p>"

KeyError: 'color'
]

Also, highlighting of coded words starts in the middle of words -- it seems things shifted a little. This applies to code reports (first and last words are sometimes truncated).

**Comment by ccbogel** [comment]
I have fixed the recognition and setting of a coder name on import.
The missing special characters - yes an issue - as this will also affect code positions - which is why you would see a shift - I reckon. Not sure how to resolve at the moment.
Key error color -  I dont get this error, but there have been a few changes made. Perhaps try to load the refi-qda file again and let me know if it still occurs. ps I have also added some more code to help catch and identify the cause.

**Comment by glocalglocal** [comment]
No error this time but the shifting/special characters issue persists. 

Also, the many codes are in groups of mostly grey, and then magenta, green and orange -- 4 colours in total.  

And the usual confusion over coder names until I got the right one to reveal underlying codings. 

BTW, I hadn't realised I can move codes to other sub-sub-categories etc or merge them with other codes. Nice!

**Comment by ccbogel** [comment]
OK- And the usual confusion over coder names until I got the right one to reveal underlying codings. - although one of the coder names from the original project should have been set there after import.
NO CODE SELECTED inherits the colour  - yes an artifact that I would has a bit of difficult resolving for now. as I dont know how to remove the fill background on the label.
The many codes in the groups of grey magenta etc - hmm - not sure ahow to resolve - these are a lot different to the nvivo original colours for each code? If you wre to unzip the qpdx file and send me the project.qde code - I might be able to look at this further - but not essential for now.

**Comment by glocalglocal** [comment]
I got in by switching off backups in the .ini as you said. I created a test nvivo export for more controlled conditions, imported it into qualcoder and it worked mostly OK. The correct coder was selected, and all the codings and codes are there. The number of different colours is correct and they are assigned OK, but qualcoder replaced the colours -- not a problem as long as they are different and it's difficult to match the colours anyway because Nvivo only supports 9 colours! 

The source of the text was qualcoder's github page. Importing the project meant that the rich text formatting of the webpage was lost. 

Also, the special characters/shift issue persists.  

I can send you the .qdpx and qualcoder files if it helps.

**Comment by ccbogel** [comment]
OK good to know some details are OK, good to kmow colours are OK also. 
Yes rich text formatting will be lost - as qualcoder only works with plain text for coding (any formatting richness is a nightmare to deal with).
Yes special characters will be a problem - I have put this in the too hard basket for now - but - please send the files and I will see if I can look at the special characters and work out a solution at some stage.
p.s. you must still likeQualCoder LOL despite all these issues

**Comment by glocalglocal** [comment]
I am attaching the test .qdpx and qualcoder files. 

I liked qualcoder before but it has improved to the point I think it can easily compete with the big guys in code and retrieve. There are still issues but you are sorting them out incredibly fast. NVivo et al do all sorts of pointless things like word clouds people used at conferences and brochure covers last century.  

I think qualcoder is a credible FOSS alternative. I haven't tried the audio/video transcribe and code features yet -- I will need to install VLC 32bit for that. But my impression is that it works pretty much in the same way pro applications work and if you can add foot pedal control, qualcoder will be better than many expensive applications like Express Scribe Pro that people can't afford. 

It's 3:30 here so that's it from me for now. 

[test.zip](https://github.com/ccbogel/QualCoder/files/4906145/test.zip)

**Comment by ccbogel** [comment]
Thanks very much for the feedback. Yes - I have not been impressed with word clouds - and it would require extra modules to import.
Wow 3.30 - have a good night and sleep in. 1pm here, lunchtime.
VLC - particularly on Windows is an issue at times.

**Comment by glocalglocal** [comment]
Promotion: I do my bit when talking to people. The documentation, support and speed of development are major pluses. The no-frills approach is good for stability, lightness and speed. The user friendliness that RDQA never had is good. The only downside is perhaps polish and standardisation in the UI, but addressing these will probably disrupt things and add weight. Eg a typical user would expect 'Recent projects' (not working at the moment) to show a sub-menu, not another window. Several windows are unnecessarily big and have too much white space. Font sizes are inconsistent here. Entering a word in a search field could be followed by enter (now you have to take your hands off the keyboard and click on the button). But as I say, these are minor things and when some of the stability/features issues are resolved, qualcoder is already a great tool for production. 

Speaking of production, will it at some point be optionally packaged for easy installation by novice users who don't want to mess with python? And will there be a 'check for updates'? A create desktop icon? Association with *.rqa files? I think packaging will be important because people don't want their qualcoder setup to stop working halfway through their project because of broken dependencies -- which is what happened with RDQA.

**Comment by ccbogel** [comment]
Thanks - yes good advice. I will gradually work on polish etc. Just have to work through the current list of issues first. (Right now looking at the dynamic code highlighting).
I dont know what an rqa file is. 

Packaging is a real pain - I have managed to create a Debian installer - although user still should install some dependencies through the terminal. I looked at flatpak - but found it too hard to follow.
I looked at pyinstaller for Windows but got stuck. 
I guess the best approach for now, is to continue fixing most issues and then getting others to assist by showing how or doing the packaging.

**Comment by glocalglocal** [comment]
I meant *.qda -- the files qualcoder creates :) It would be nice and reassuring to be able to double-click a qda file in Windows Explorer (complete with an icon and Type description) to open it in qualcoder.  

The reason I mentioned packaging is because I saw recently what happened with people (inc myself) desperately trying to get RQDA to work in R v4.0.2 and RStudio with dependencies dropped overnight.

**Comment by ccbogel** [comment]
yes sure. the packaging is beyond me unfortunately. I got very frustrated trying ti make it work.

**Comment by glocalglocal** [comment]
> Also, the special characters/shift issue persists.

This also affects the tooltip text when hovering over coded text.

**Comment by ccbogel** [comment]
I made a small change to the refi-qda.qdpx import of text files. I tested with your test.qdpx
I did not see any character shifting. So I am hoping this is a success. Picture below:
![pic1](https://user-images.githubusercontent.com/5834638/89403641-f6c01280-d75b-11ea-812d-025da0423ed1.png)

**Comment by glocalglocal** [comment]
This file works but I imported a complete project with 73 files. Manage Files/Open and Coding Reports show codings correctly but Code Text shows no codings at all -- ie codings are not displayed in the text box or in the combo.

**Comment by ccbogel** [comment]
Is this because the coder name is different?
Check the coder names available in the Settings. It might be that the project has 2 codernames and the one that did not do any/much coding has been selected.

**Comment by glocalglocal** [comment]
Yes, it was the coder name. My stupid mistake (again), but I am thinking if I make this mistake repeatedly despite having spent so much time on qualcoder, others may get confused too. Perhaps dialogues should indicate the list of coders? 

Otherwise, funny characters are gone and shifting is almost resolved. Every paragraph now starts with a space -- seen also in your screenshot above. This causes a shift by one character. Codings start and end one character early.

**Comment by ccbogel** [comment]
I prefer not to have the coder name selected apart from in Settings. This is because when dual coding text, it would be so easy to use the wrong name to code over the text. Maybe it would be better to let the user know that if no codings are present, then they should check and change the codername in Settings.
Glad shifting is almost resolved, it seemed to be a challenging problem to me.

**Comment by glocalglocal** [comment]
> Maybe it would be better to let the user know that if no codings are present, then they should check and change the codername in Settings.

Makes sense to me.

**Comment by ccbogel** [comment]
Yes. In the same way I added  a dialog to explain the 2 steps to import rrqda and qdpx files

**Comment by ccbogel** [comment]
added the message
